src/lxc/attach.c

```diff
@@ -199,52 +201,60 @@ static int in_same_namespace(pid_t pid1, pid_t pid2, const char *ns)
 		if (errno == ENOENT)
 			return -EINVAL;
 
+		saved_errno = errno;
 		goto out;
 	}
 
 	ns_fd2 = lxc_preserve_ns(pid2, ns);
-	if (ns_fd2 < 0)
+	if (ns_fd2 < 0) {
+		saved_errno = errno;
 		goto out;
+	}
 
 	ret = fstat(ns_fd1, &ns_st1);
-	if (ret < 0)
+	if (ret < 0) {
+		saved_errno = errno;
 		goto out;
+	}
 
 	ret = fstat(ns_fd2, &ns_st2);
-	if (ret < 0)
+	if (ret < 0) {
+		saved_errno = errno;
 		goto out;
+	}
 
 	/* processes are in the same namespace */
-	ret = -EINVAL;
-	if ((ns_st1.st_dev == ns_st2.st_dev ) && (ns_st1.st_ino == ns_st2.st_ino))
+	if ((ns_st1.st_dev == ns_st2.st_dev ) && (ns_st1.st_ino == ns_st2.st_ino)) {
+		ret = -EINVAL;
 		goto out;
+	}
 
 	/* processes are in different namespaces */
 	ret = ns_fd2;
 	ns_fd2 = -1;
 
 out:
-
 	if (ns_fd1 >= 0)
 		close(ns_fd1);
+
 	if (ns_fd2 >= 0)
 		close(ns_fd2);
 
+	errno = saved_errno;
 	return ret;
 }
 
 static int lxc_attach_to_ns(pid_t pid, struct lxc_proc_context_info *ctx)
 {
-	int i, ret;
+	int i;
 
 	for (i = 0; i < LXC_NS_MAX; i++) {
 		if (ctx->ns_fd[i] < 0)
 			continue;
 
-		ret = setns(ctx->ns_fd[i], ns_info[i].clone_flag);
-		if (ret < 0) {
+		if (setns(ctx->ns_fd[i], ns_info[i].clone_flag) < 0) {
```

Comment:
According to [1] the old version was the preferred way in this case. @brauner should have a stronger opinion here. We do have both in the code base.
[1] https://github.com/lxc/lxc/blob/master/CODING_STYLE.md#functions-not-returning-booleans-must-assign-return-value-before-performing-checks

Reply:
Yes, assigning is the preferred way. :)

Reply:
The ret variable is assigned to stack area. but it is used only 1 time in function. So, I thought it is inefficient because a return value is stored in register and use it directly. But this is the coding rule, I'll change it. :)
Thanks for your review.

Reply:
Yeah, these are corner cases so I won't really back it based on that but it's preferred in general. So you can leave it if you want. :)